### PR TITLE
Implement coroutines interop module

### LINKED ISSUE #371: ScopeProvider -> CoroutineScope interop
Can we do something like this?

```kotlin
fun CoroutineScope.asScopeProvider(): ScopeProvider

fun ScopeProvider.asCoroutineScope(): CoroutineScope

fun CoroutineScope.asCompletable(): Completable

fun Completable.asCoroutineScope(): CoroutineScope
```

### PR DESCRIPTION
**Pull request by ZacSweers**
This introduces a new interop artifact for interop between `Completable`/`ScopeProvider` and `CoroutineScope`. The idea is to offer this interop for general use, but also so anyone migrating from rx to coroutines (or the other way around!) can reuse existing machinery to ease migrations in the meantime.

Resolves #371

### CODE REVIEW

File: autodispose-coroutines-interop/src/main/kotlin/com/uber/autodispose/coroutinesinterop/AutoDisposeCoroutinesInterop.kt [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+ *         when [this] scope provider completes.
+ */
+fun CompletableSource.asCoroutineScope(context: CoroutineContext): CoroutineScope {
+  val scope = CoroutineScope(SupervisorJob() + context)
+
+  // Bind to the scope, so if the scope is manually canceled before our scope provider emits, we
+  // clean up here.
+  Completable.wrap(this)
+      .autoDispose(scope)
+      .subscribe({ scope.cancel() }) { e -> scope.cancel("OnError", e) }
```
[the remark this answers is not in the document]
Comment (ZacSweers) [reply autodispose-coroutines-interop/src/main/kotlin/com/uber/autodispose/coroutinesinterop/AutoDisposeCoroutinesInterop.kt]: I like using the trailing lambda when possible!

**Comment by ZacSweers** [comment]
Implemented the tests in d186f8f

I'm still not totally sure about requiring a `CoroutineContext`, vs maybe giving it a default value of something like `Job()` or `SupervisorJob()`. Thoughts?

**Comment by ZacSweers** [comment]
Went for it after digging through some coroutines sources. I think that's the right move.

**Comment by ZacSweers** [comment]
Wrote something really nice building on top of the building blocks here. Now you can do this:

```kotlin
scopeProvider.asCoroutineScope {
  someObservable.autoDispose().subscribe()
}
```

**Comment by ZacSweers** [comment]
Pulled out the context APIs to #375 instead

File: autodispose-coroutines-interop/src/test/kotlin/com/uber/autodispose/coroutinesintrop/AutoDisposeCoroutinesInteropTest.kt [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+import io.reactivex.subjects.PublishSubject
+import io.reactivex.subjects.SingleSubject
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.ensureActive
+import kotlinx.coroutines.isActive
+import org.junit.Test
+
+class AutoDisposeCoroutinesInteropTest {
```
Comment (ShaishavGandhi) [inline autodispose-coroutines-interop/src/test/kotlin/com/uber/autodispose/coroutinesintrop/AutoDisposeCoroutinesInteropTest.kt]: Nice tests!

**Pull request merged**